ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/batch/ConflictDetection.scala

```diff
@@ -0,0 +1,150 @@
+        .hashCode() || inputValue != outputValue
+      if (!contentsDiffer) {
+        logger.trace(s"Dropped key=$key from conflicting key set as its value has not been altered")
+        Metrics.removedTransientKey.inc()
+      }
+      contentsDiffer
+    }
+
+  private def produceRejectionOrDropSubmission(
+      logEntry: DamlLogEntry,
+      conflictingKeys: Set[DamlStateKey])(implicit loggingContext: LoggingContext)
+    : Option[(DamlLogEntry, Map[DamlStateKey, DamlStateValue])] = {
+    Metrics.conflicted.inc()
+
+    logEntry.getPayloadCase match {
+      case TRANSACTION_ENTRY =>
+        Metrics.recovered.inc()
+        val reason = explainConflict(conflictingKeys)
+        val transactionRejectionEntry = transactionRejectionEntryFrom(logEntry, reason)
+        logger.trace(s"Recovered a conflicted transaction, details='$reason'")
+        Some((transactionRejectionEntry, Map.empty))
+
+      case PARTY_ALLOCATION_ENTRY | PACKAGE_UPLOAD_ENTRY | CONFIGURATION_ENTRY |
+          TRANSACTION_REJECTION_ENTRY | CONFIGURATION_REJECTION_ENTRY |
+          PACKAGE_UPLOAD_REJECTION_ENTRY | PARTY_ALLOCATION_REJECTION_ENTRY =>
+        logger.trace(s"Dropping conflicting submission (${logEntry.getPayloadCase})")
+        Metrics.dropped.inc()
+        None
+
+      case PAYLOAD_NOT_SET =>
+        sys.error("conflictDetectAndRecover: PAYLOAD_NOT_SET")
+    }
+  }
+
+  // Attempt to produce a useful message by collecting the first conflicting
+  // contract id or contract key.
+  private def explainConflict(conflictingKeys: Iterable[DamlStateKey]): String =
+    conflictingKeys
+      .collectFirst {
+        case key if key.hasContractKey =>
+          // NOTE(JM): We show the template id as the other piece of data we have is the contract key
+          // hash, which isn't very useful as it's an implementation detail not exposed over ledger-api.
+          val templateId =
+            ValueCoder
+              .decodeIdentifier(key.getContractKey.getTemplateId)
+              .map(_.toString)
+              .getOrElse("<unknown>")
+          s"Contract key conflicts in contract template $templateId"
+
+        case key if key.getContractId.nonEmpty =>
+          s"Conflict on contract ${key.getContractId}"
+
+        case key if key.hasCommandDedup =>
+          "Conflict on command deduplication, this command has already been processed."
+
+        case key if key.hasConfiguration =>
+          "Ledger configuration has changed"
+      }
+      .getOrElse("Unspecified conflict")
+
+  private def transactionRejectionEntryFrom(
+      logEntry: DamlLogEntry,
+      reason: String): DamlLogEntry = {
+    val builder = DamlLogEntry.newBuilder
+    builder.setRecordTime(logEntry.getRecordTime)
+    builder.getTransactionRejectionEntryBuilder
+      .setSubmitterInfo(logEntry.getTransactionEntry.getSubmitterInfo)
+      .getInconsistentBuilder
+      .setDetails(reason)
+    builder.build
+  }
+
+  private[batch] object Metrics {
```

Comment:
Please move the metrics to `ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala`

Reply:
That's why it's a bad idea to centralize metrics, i.e., now the whole world needs to depend on that single file.